### PR TITLE
Propagate arguments

### LINKED ISSUE #526: Overload resolution for .value()-modified functions
[I have](https://github.com/axic/shapeshiftbot) a method, which gets called with value transfer using the `<method>.value(<value>)(<args>)` semantic. This fails when overloading.

Works:

```
contract ShapeshiftBot {
  function transfer(string coin, string recipient, bool acceptReturn, address returnAddress) returns (bytes32 myid);
  function transfer(string coin, string recipient, bool acceptReturn) returns (bytes32 myid);
  function transfer(string coin, string recipient) returns (bytes32 myid);
}

contract ShapeshiftBotLookup {
  function getAddress() returns (address _addr);
}

contract usingShapeshift {
  function shapeshiftInstance() returns (ShapeshiftBot) {
    ShapeshiftBotLookup lookup = ShapeshiftBotLookup(0x0);
    return ShapeshiftBot(lookup.getAddress());
  }

  function shapeshiftTransfer(uint value, string coin, string recipient, bool acceptReturn, address returnAddress) internal returns (bytes32 myid) {
    return shapeshiftInstance().transfer(coin, recipient, acceptReturn, returnAddress);
  }

  function shapeshiftTransfer(uint value, string coin, string recipient, bool acceptReturn) internal returns (bytes32 myid) {
    return shapeshiftInstance().transfer(coin, recipient, acceptReturn);
  }

  function shapeshiftTransfer(uint value, string coin, string recipient) internal returns (bytes32 myid) {
    return shapeshiftInstance().transfer(coin, recipient);
  }
}
```

Fails when adding `.value()` in `usingShapeshift`:

```
Error: Member "transfer" not unique after argument-dependent lookup in contract ShapeshiftBot
    return shapeshiftInstance().transfer.value(value)(coin, recipient, acceptReturn, returnAddress);
           ^---------------------------^
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/526

### CODE REVIEW

**Comment by axic** [comment]
Can you add a test case for `{gas:..}` too? And perhaps a combined value and gas case?

**Comment by axic** [comment]
~~Issue #3556 had a comment that perhaps that can also be solved by the same change. The change you made here suggests otherwise, but can you check?~~

Never mind, read it again, not sure if it can be solved without introducing some new syntax.

**Comment by chriseth** [comment]
Also turned it into a semantic test.

File: test/libsolidity/semanticTests/functionCall/call_options_overload.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+contract C {
+    function f(uint x) external payable returns (uint) { return 1; }
+    function f(uint x, uint y) external payable returns (uint) { return 2; }
+    function call() public payable returns (uint x, uint y) {
+        x = this.f{value: 10}(2);
+        y = this.f{gas: 1000}(2, 3);
+    }
```
Comment (axic) [inline test/libsolidity/semanticTests/functionCall/call_options_overload.sol:9]: Can you also add a mixed case (`f{value:10, gas:100}`) ?

Comment (chriseth) [reply test/libsolidity/semanticTests/functionCall/call_options_overload.sol:9]: oh come on - tests are always written keeping the actual implementation in mind - if you just randomly test stuff it won't help you anyway and the implementation obviously shows that you can have anything inside there even `this.f{value: f(2)}(2)`.

Comment (axic) [reply test/libsolidity/semanticTests/functionCall/call_options_overload.sol:9]: Is it a sane approach to only consider the implementation and not what people may end up doing?

**Review by axic — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -2354,6 +2354,8 @@ bool TypeChecker::visit(FunctionCallOptions const& _functionCallOptions)
 {
 	solAssert(_functionCallOptions.options().size() == _functionCallOptions.names().size(), "Lengths of name & value arrays differ!");
 
+	_functionCallOptions.expression().annotation().arguments = _functionCallOptions.annotation().arguments;
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:2357]: Oh wow, that's it?!

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**